### PR TITLE
htc #20379 allow registering custom media types

### LINKED ISSUE #20379: Allow registering custom MediaTypes, and do it non-globally
In Spray this was done by exposing the `register()` method on `MediaTypes` however this has the downside of being global state – we must register custom types scoped to an ActorSystem.

### PR DESCRIPTION
**Pull request by ktoso**
Resolves #20379

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/javadsl/settings/ParserSettings.scala [hunk 1]
```diff
@@ -68,6 +69,11 @@ abstract class ParserSettings private[akka] () extends BodyPartParser.Settings {
     val map = codes.map(c ⇒ c.intValue -> c.asScala).toMap
     self.copy(customStatusCodes = map.get)
   }
+  @varargs
+  def withCustomMediaTypes(mediaTypes: MediaType*): ParserSettings = {
+    val map = mediaTypes.map(c ⇒ (c.mainType, c.subType) -> c.asScala).toMap
+    self.copy(customMediaTypes = (main, sub) ⇒ map.get(main, sub))
```
Comment (patriknw) [inline akka-http-core/src/main/scala/akka/http/javadsl/settings/ParserSettings.scala]: is this auto-tupling? shouldn't that be avoided? `map.get((main, sub))`

File: akka-http-tests/src/test/resources/httpsDemoKeys/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# expects the password to be "abcdef"):
+
+```
+openssl pkcs12 -export -name servercrt -in chain.pem -inkey server.key -out server.p12
+```
+
+# For investigating remote certs use:
+
+```
+openssl s_client -showcerts -connect 54.173.126.144:443
```
Comment (patriknw) [inline akka-http-tests/src/test/resources/httpsDemoKeys/README.md]: looks very unrelated

**Comment by patriknw** [comment]
LGTM

File: akka-http-core/src/main/scala/akka/http/javadsl/settings/ParserSettings.scala (hunk 1 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/settings/ParserSettings.scala]: Whoops, good catch

**Comment by ktoso** [comment]
Rebased and removed files from other PR which sneaked in here

File: akka-http-tests/src/test/resources/httpsDemoKeys/README.md (hunk 2 above)
Comment (ktoso) [inline akka-http-tests/src/test/resources/httpsDemoKeys/README.md]: removed

File: akka-http-tests/src/test/scala/akka/http/scaladsl/CustomMediaTypesSpec.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
+
+      //#application-custom
+
+      // similarily in Java: `akka.http.javadsl.settings.[...]`
+      import akka.http.scaladsl.settings.ParserSettings
+      import akka.http.scaladsl.settings.ServerSettings
+
+      // define custom media type:
+      val utf8 = HttpCharsets.`UTF-8`
```
Comment (drewhk) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/CustomMediaTypesSpec.scala]: why does this need to be stored in a val?

Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/CustomMediaTypesSpec.scala]: only in order to make the line below fit in a single line when included into docs

**Comment by ktoso** [comment]
Failure was Mima.
Added MiMa filters (they're safe, internal settings changes) and will merge after validation.

**Pull request merged**